### ISSUE TITLE
Multi Word Facet issue

### ISSUE THREAD

**Issue by divyaiyer** [karmi/retire#67]
I am incorporating Tire gem to one of my rails application. I have a derived field called channel_name which I need to use for faceting. My model looks like this

  include Tire::Model::Search
  include Tire::Model::Callbacks

 mapping do
    indexes :channel_name,:type => 'string',:index=> :not_analyzed
  end

   def to_indexed_json
        channel_name =self.group_channel.name

```
    {
      :title => title,
      :description => description,
      :channel_name =>channel_name,
      :created_at=>created_at
    }.to_json
```

   end

And I build index like this Show.import :per_page => 500

And when I search using this
 q = 'series'
    s=Show.search do
      query { string q }
      facet('channel_name') { terms :channel_name}
      sort { created_at 'desc' }
    end

I am getting facets as single words. Ex: I have a channel name 'This is India' , but facets comes as

 term=>'this',count=>2
 term=>'is',count=>2
 term=>'India',count=>2

But I need this term=>'this is India',count=>2

Could you please tell me where I am going wrong. I checked the index created in data folder of elastic search , but there my channel_name got indexed properly.

**Comment by karmi** [comment]
Hi,

first: your mapping definition is correct. You need to make absolutely sure, that this is the mapping in ES. Either call `Show.elasticsearch_index.mapping` or just load http://localhost:9200/shows/_mapping in curl/browser. By the results the facet give you back, the `channel_name` is, in fact analyzed.

Take a look at & run this code, it gives proper facet back:

``` ruby
    require 'rubygems'
    require 'active_record'
    require 'tire'

    ActiveRecord::Base.establish_connection( :adapter => 'sqlite3', :database => ":memory:" )

    ActiveRecord::Migration.verbose = false
    ActiveRecord::Schema.define(:version => 1) do
      create_table :shows do |t|
        t.string   :title, :description, :channel_name
        t.datetime :created_at, :default => 'NOW()'
      end
    end

    # 1) Make sure the index does not exist, so it is created with proper mapping
    #    (See section about FORCE importing in the Readme)
    Tire.index('shows').delete

    class Show < ActiveRecord::Base
      include Tire::Model::Search
      include Tire::Model::Callbacks


      mapping do
        indexes :channel_name, :type => 'string', :index => :not_analyzed
      end

      # 2) Provide your own, custom `to_indexed_json` method
      def to_indexed_json
        {
          :title        => title,
          :description  => description,
          :channel_name => channel_name,
          :created_at   => created_at
        }.to_json
      end
    end

    # 3) Create a document and refresh the index
    Show.create :title => 'Series about ...', :channel_name => 'This is India'
    Show.elasticsearch_index.refresh

    s = Show.search do
      query { string 'series' }
      facet('channel_name') { terms :channel_name}
      sort { by :created_at, 'desc' }
    end

    p s.facets
    # =>
    # {
    #     "channel_name"=>
    #     {
    #         "total"=>1,
    #          "_type"=>"terms",
    #          "other"=>0,
    #          "terms"=>
    #         [
    #             {
    #                 "term"=>"This is India",
    #                  "count"=>1
    #             }
    #         ],
    #          "missing"=>0
    #     }
    # }
```

**Comment by divyaiyer** [comment]
I will try this again from scratch. Mapping of shows gives channel_name type
as String. Its weird why I am getting like this. I shall try your code on a
new model & see, whether that works for me.

On Wed, Aug 3, 2011 at 7:08 PM, karmi <
reply@reply.github.com>wrote:

> Hi,
> 
> first: your mapping definition is correct. You need to make absolutely
> sure, that this is the mapping in ES. Either call
> `Show.elasticsearch_index.mapping` or just load <
> http://localhost:9200/shows/_mapping> in curl/browser. By the results the
> facet give you back, the `channel_name` is, in fact analyzed.
> 
> Take a look at & run this code, it gives proper facet back:
> 
> ``` ruby
>    require 'rubygems'
>    require 'active_record'
>    require 'tire'
> 
>    ActiveRecord::Base.establish_connection( :adapter => 'sqlite3',
> :database => ":memory:" )
> 
>    ActiveRecord::Migration.verbose = false
>    ActiveRecord::Schema.define(:version => 1) do
>      create_table :shows do |t|
>        t.string   :title, :description, :channel_name
>        t.datetime :created_at, :default => 'NOW()'
>      end
>    end
> 
>    # 1) Make sure the index does not exist, so it is created with proper
> mapping
>    #    (See section about FORCE importing in the Readme)
>    Tire.index('shows').delete
> 
>    class Show < ActiveRecord::Base
>      include Tire::Model::Search
>      include Tire::Model::Callbacks
> 
> 
>      mapping do
>        indexes :channel_name, :type => 'string', :index => :not_analyzed
>      end
> 
>      # 2) Provide your own, custom `to_indexed_json` method
>      def to_indexed_json
>        {
>          :title        => title,
>          :description  => description,
>          :channel_name => channel_name,
>          :created_at   => created_at
>        }.to_json
>      end
>    end
> 
>    # 3) Create a document and refresh the index
>    Show.create :title => 'Series about ...', :channel_name => 'This is
> India'
>    Show.elasticsearch_index.refresh
> 
>    s = Show.search do
>      query { string 'series' }
>      facet('channel_name') { terms :channel_name}
>      sort { by :created_at, 'desc' }
>    end
> 
>    p s.facets
>    # =>
>    # {
>    #     "channel_name"=>
>    #     {
>    #         "total"=>1,
>    #          "_type"=>"terms",
>    #          "other"=>0,
>    #          "terms"=>
>    #         [
>    #             {
>    #                 "term"=>"This is India",
>    #                  "count"=>1
>    #             }
>    #         ],
>    #          "missing"=>0
>    #     }
>    # }
> ```
> 
> ## 
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/karmi/tire/issues/67#issuecomment-1718006

## 

Regards
Divya Shyam

**Comment by karmi** [comment]
Believe me, the most probable cause for this is that the index is already created _before_ your model is loaded. Delete the index (`curl -X DELETE http://localhost:9200/shows`), and load the app. Or use the `FORCE` option in the Rake task, as documented in the Readme.

**Comment by divyaiyer** [comment]
For my surprise it worked for a replica of that app on my Laptop. I
installed the same-thing on my lappie. And now when I run search I get it
fixed. But really strange and I dont know what was the issue and hence how
to fix the issue on my office comp. Anyways thanks a lot for your quick
responses. Lemme try to figure out, what was the bug behind it.

Cheers
Div

**Comment by karmi** [comment]
Again, something must be creating that index before your app is loaded...

**Comment by divyaiyer** [comment]
Yeah, whether it cud be this? I have two models and using same attribute
channel_name and mapping in both the models. Bcz for the search  I need
objects of both models.

On Wed, Aug 3, 2011 at 9:51 PM, karmi <
reply@reply.github.com>wrote:

> Again, something must be creating that index before your app is loaded...
> 
> ## 
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/karmi/tire/issues/67#issuecomment-1719344

## 

Regards
Divya Shyam

**Comment by karmi** [comment]
No, different models map to different indices in ES.

**Comment by karmi** [comment]
So, any luck in the end?

**Comment by divyaiyer** [comment]
Unfortunately no. Did re-installation elastic server itself and removed all
model Tire initialization. And sample code executed in rails console. But result
was same at office PC.

On Tue, Aug 16, 2011 at 7:18 PM, karmi <
reply@reply.github.com>wrote:

> So, any luck in the end?
> 
> ## 
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/karmi/tire/issues/67#issuecomment-1816341

## 

Regards
Divya Shyam

**Comment by karmi** [comment]
Start with a fresh app, and copy over only the models & migrations etc. If the problem persist, try to start with a fresh app with a simple model, hooked to Tire, and submit a new issue, please...

**Comment by divyaiyer** [comment]
Sure, off for the day. Wud try tomorrow and submit the issue if persists .

On Tue, Aug 16, 2011 at 9:48 PM, karmi <
reply@reply.github.com>wrote:

> Start with a fresh app, and copy over only the models & migrations etc. If
> the problem persist, try to start with a fresh app with a simple model,
> hooked to Tire, and submit a new issue, please...
> 
> ## 
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/karmi/tire/issues/67#issuecomment-1817863

## 

Regards
Divya Shyam

**Comment by oleksify** [comment]
Can somebody help me with this issue, I just can't resolve it. I re index data, delete index, restart servers, everything.

my mappings (i tried a lot of other combinations)

  tire.mapping do
    indexes :product_id,         :type => 'integer', :index    => :not_analyzed
    indexes :name,               :type => 'string',  :analyzer => 'russian'
    indexes :deleted,            :type => 'boolean'
    indexes :short_description,  :type => 'string',  :analyzer => 'russian'
    indexes :original_currency,  :type => 'integer'
    indexes :brand,              :type => 'string',  :index    => 'not_analyzed'
    indexes :url,                :type => 'string',  :index    => 'not_analyzed'
    indexes :price,              :type => 'float'
    indexes :taxons,             :type => 'integer', :analyzer => 'keywords'
    indexes :facets,             :type => 'nested',  :index    => 'not_analyzed'
    indexes :images,             :type => 'string',  :index    => 'not_analyzed'
  end 

facets is a Hash of values. 

later on in controller i use:

  facets.each do |f|
    facet f.name do
      terms f.name.to_sym, :size => Product.not_deleted.count
    end
  end

and it counts perfect, everything perfect but it shows my facets splited on words, so facet "250 kVt" is now 2 facets "250" & "kVt"

Please, help me!

**Comment by karmi** [comment]
@devilcoders, you have to set the `name` field to be `not_analyzed`. Search "multi field" on the ES site...

**Comment by oleksify** [comment]
name field in mapping is name of the product, not related to facets.  I need to define facets dynamicly from darabase. so I get facets array (other model object from database) and place it this way as I wrote:
<code>
  facets.each do |f|
    facet f.name do
      terms f.name.to_sym, :size => Product.not_deleted.count
    end
  end
</code>

Right jnow I update elasticsearch to 18.2 and now it stopws giving me any facets from nested document. What should I do?

**Comment by oleksify** [comment]
Sorry, I fugure everything out already, it was my silly mistake. Everything works great right now. Thanks!

**Comment by xdmx** [comment]
Hi Karmi, what about multi_fields under Tire that needs to be both analyzed and not?

I've tried something like this (plus many others combinations):

indexes :tags, :type => 'multi_field', :index_name => 'tag', :fields => {
  :tag       => { :type => 'string', :analyzer => 'standard' },
  :untouched => { :type => 'string', :analyzer => 'not_analyzed', :include_in_all => false }
}

But then after I index everything I don't get (I check under es-head) any tag field, just tags... the strange behavior is also that after I add that line also 

```
indexes :languages, :type => 'string', :index_name => 'language', :index => 'not_analyzed', :include_in_all => false
```

(is a string array, like ['en', 'de', 'fr']) stops to work and it just shows languages and not language anymore

but if I use 

```
indexes :tags, :type => 'string', :index_name => 'tag', :include_in_all => false
```

then they both work

But unfortunately I need to analyze the tags in order to do searches on them, but show the whole tag on facets and not just tokens

Then how make the facet call on the untouched one?

  facet('tags') { terms 'tag.untouched' } ?

Thanks!

**Comment by karmi** [comment]
@xdmx: I'm not sure I understand the problem 100% :), but setting the proper index in Tire is easy, eg.:

``` ruby
class MyModel < CouchRest::Model::Base
  include Tire::Model::Search
  include Tire::Model::Callbacks

  mapping do
    indexes :author,
              :type => 'multi_field',
              :fields => {
                :author => { :type => 'string', :analyzer => 'czech', :boost => 10.0 },
                :exact  => { :type => 'string', :analyzer => 'stop',   :boost => 10.0 }
              }
    end
  end

end
```

The facet definition is again easy:

``` ruby
facet('authors')   { terms "author.exact", :size => 5 }
```

Is this what you're after?

**Comment by xdmx** [comment]
mmm..I try to give you an example :)

```
class Post < ActiveRecord::Base
  include Tire::Model::Search
  include Tire::Model::Callbacks

  mapping do
indexes :languages, :type => 'string', :index_name => 'language', :index => 'not_analyzed', :include_in_all => false
indexes :tags, :type => 'multi_field', :fields => { :tags      => { :type => 'string', :index_name => 'tag', :analyzer => 'standard' },
                            :untouched => { :type => 'string', :analyzer => 'not_analyzed', :include_in_all => false } }
  end

  def to_indexed_json
{:languages => ['it', 'en'], :tags => ['First tag', 'Second tag']}.to_json
  end
end

Post.search do
  query { string 'first' }
  facet('tags')   { terms "tags" }
end # => facets['tags']['total'] => 259743 ## ok, all the various tokens: "first", "second", "tag".... 86581 records * 3

Post.search do
  query { string 'first' }
  facet('tags')   { terms "tags.untouched" }
end # => facets['tags']['total'] => 0 ## but it should be 86581 for "First tag"!
```

I've tried also to do some searches:

```
Post.search do
  query { string 'tag:first' }
end # => total => 0 ## what? ignored the index_name?

Post.search do
  query { string 'tags:first' }
end # => total => 86581 ## ok

Post.search do
  query { string 'language:it' }
end # => total => 0 # seems ignored too

Post.search do
  query { string 'languages:it' }
end # => total => 0 # ok, that's strange too

Post.search do
  query { string 'tags.tags:first' }
end # => total => 0 ## mmmmmmmm

Post.search do
  query { string 'tags.tag:first' }
end # => total => 0 ## mmmmmmmm
```

I guess that "tag" and "language" should be recognized as there is the index_name.... looking the index from es-head there is just languages and tags, nothing else

**Comment by xdmx** [comment]
Just an update, it was actually my fault, I wrote :analyzer => 'not_analyzed', instead of :index => 'not_analyzed'... and that created just a very weird behavior. I think it would be better to ignore that not supported cases, in my project I had duplicated indexes for array values (eg. language and languages), without be able to search in any of them

**Comment by karmi** [comment]
@xdmx: Good, it was suspicious all the time. Thanks for the suggestion, but _Tire_ does not want to interfere with user's possibly superior knowledge. Garbage in, garbage out, good stuff in, good stuff out...

**Comment by christianhellsten** [comment]
This issue seems to pop up every now and then. Usually it goes away after deleting the index and reindexing: 
- I have checked that the field is not analyzed:
  
  ```
  indexes :category, :type => 'string', :index => :not_analyzed
  ```
- I have double checked the mapping and ":index => :not_analyzed" is missing
  
  ```
  http://localhost:9200/products/_mapping =
  {"products":{"product":{"properties":{"category":{"type":"string"}}}}
  ```
- I have checked tire.mapping and tire.index.mapping:
  
  ```
  Product.tire.mapping
  => {:category=>{:index=>:not_analyzed, :type=>"string"}}
  
  Product.tire.index.mapping
  => {"product"=>{"category"=>{"type"=>"string"}}}
  ```
- Index changes between console sessions
  
  ```
  # First time
  $b rails console      
  
  > Product.tire.index.mapping
  => {"product"=>{"properties"=>{"category"=>{"type"=>"string", "index"=>"not_analyzed"}}}}
  > exit
  
  # Immediately after
  $b rails console      
  
  # Product.tire.index.mapping
  => {"product"=>{"properties"=>{"category"=>{"type"=>"string"}}}}
  ```
- This is how the facet is defined:
  
  ```
  facet 'category' do
    terms :category, :global => true
  end
  ```

**Comment by christianhellsten** [comment]
OK, I put a debug statement in Tire::Model::Indexing::ClassMethods#create_elasticsearch_index. Found out that this Rake task caused the issue:

```
namespace :search do
  desc "Reindex"
  task :reindex => :environment do
    Tire.index('products').delete
    Product.import :per_page => 1000
  end
end
```

If I run this Rake task, the mapping is not what I have defined in the model. I checked http://localhost:9200/products/_mapping.

I noticed Tire::Model::Indexing::ClassMethods#create_elasticsearch_index is run only once when Tire.index('products').delete is called. This means the mapping is empty when the import statement is run and I guess ElasticSearch's default behavior is to create the mapping from whatever data is in the bulk import.

Is there something we can do to avoid this?

**Comment by karmi** [comment]
@christianhellsten, I'm sorry but I got lost in what problem are you experiencing exactly.

First, yes, the index is created on class load, _only_ when it does _not_ exist, the Readme says so.

Second, the Rake task you're posting seems suspicious. Notice that you delete the index, **after the class has been loaded**, effectively nuking the mapping, and _then_ you import data into it. Not a good way. And yes, ES will create dynamic mapping in this case.

_Tire_ provides kinda nice import task already, if there's any issue with it, we should solve it. You can wrap this task in another task, or rip the code, etc. Feel free to ping me on the `#elasticsearch` IRC channel for more interactive help.

**Comment by christianhellsten** [comment]
Thanks. I understand the issue now and how to resolve it.

I'm just wondering if this is something that could be improved on? I don't know if it's a good idea, but Tire could, for example, detect if the Tire mapping is different from the one defined in ElasticSearch (http://localhost:9200/xxx/_mapping) and warn the user.

**Comment by karmi** [comment]
That's a very good suggestion, actually! I refrained from updating the mapping in the model integration, but this may be a very nice first step.
